feat: Added MonthylyView and tests

Roheeth05 opened the pull request:
- Displays evaluations in a grid-based calendar layout for a full month.

Tests:
- Renders weekday headers.
- Displays correct number of days (e.g., 30 for June).
- Shows evaluations on correct days.
- Accounts for padding cells before start of month.

achow-conestogac commented on src/EvaluationCalendar/MonthlyView/MonthlyView.tsx:
```diff
@@ -0,0 +1,73 @@
+              <div className="font-semibold">{date.getDate()}</div>
+            )}
+            <ul className="mt-1 space-y-1">
+              {evaluations.map((ev, i) => (
+                <li key={i} className="bg-gray-100 p-1 rounded">
+                  <div className="font-medium">{ev.title}</div>
+                  <div className="text-gray-600">{ev.course}</div>
+                </li>
+              ))}
```
what happens if evaluations is null, empty, or returns nothing?

achow-conestogac commented on src/EvaluationCalendar/MonthlyView/MonthlyView.tsx:
```diff
@@ -0,0 +1,73 @@
+      </div>
+
+      <div className="grid grid-cols-7 gap-2">
+        {calendarCells.map(({ date, evaluations }, idx) => (
```
what happens if calendarCells is null, empty, or returns nothing? 

Does the user just get a blank screen?

achow-conestogac replied:
How is this addressed?

Roheeth05 replied:
This is now handled by a fallback message inside the CalendarDayCard component.
If evaluations.length === 0, it shows:
“No evaluations scheduled for this day.”

achow-conestogac replied:
no this does not.

this is because that element will not be rendered if the `calendarCell` does not contain any element. You would have known this if TDD is followed and you have tested this situation.

furthermore, even if this is the case, you do not want to rely on another element to deal with your own error handling.  when you do that, you will end up with errors because the other component has changed.

you will need to implement your own error logic and write tests.

achow-conestogac commented on src/EvaluationCalendar/MonthlyView/MonthlyView.tsx:
```diff
@@ -0,0 +1,73 @@
+  const firstWeekday = firstDayOfMonth.getDay();
+  const daysInMonth = new Date(year, month + 1, 0).getDate();
+
+  const calendarCells: { date: Date; evaluations: Evaluation[] }[] = [];
+
+  // Empty cells before the 1st day of the month
+  for (let i = 0; i < firstWeekday; i++) {
+    calendarCells.push({ date: new Date(NaN), evaluations: [] });
+  }
+
+  for (let day = 1; day <= daysInMonth; day++) {
+    const currentDate = new Date(year, month, day);
+    const dateKey = currentDate.toDateString();
+    const dayEvaluations = evaluations.filter(
+      (ev) => ev.dueDate.toDateString() === dateKey
+    );
+    calendarCells.push({ date: currentDate, evaluations: dayEvaluations });
+  }
```
this calculation does not need to be made every single time the component refreshes. please use `useMemo` to only recalculate this if the input data changes.

the official documentation is here: https://react.dev/reference/react/useMemo

Roheeth05 replied:
Used useMemo

achow-conestogac commented on src/EvaluationCalendar/MonthlyView/MonthlyView.tsx:
```diff
@@ -0,0 +1,73 @@
+const daysInWeek = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
+
+const MonthlyView: React.FC<MonthlyViewProps> = ({ evaluations, month, year }) => {
+  const firstDayOfMonth = new Date(year, month, 1);
+  const firstWeekday = firstDayOfMonth.getDay();
+  const daysInMonth = new Date(year, month + 1, 0).getDate();
```
These variables are the same as below.  It does not need to be recalculated unless the input changes.

memoize them as well please.

Roheeth05 commented on src/EvaluationCalendar/MonthlyView/MonthlyView.test.tsx:
```diff
@@ -0,0 +1,47 @@
+
+  it("renders correct number of days in the month", () => {
+    render(<MonthlyView evaluations={[]} month={5} year={2025} />);
+    // June 2025 has 30 days
```
Removed the comments

Roheeth05 commented on src/EvaluationCalendar/MonthlyView/MonthlyView.test.tsx:
Removed the comments, Professor

Roheeth05 commented on src/EvaluationCalendar/MonthlyView/MonthlyView.tsx:
```diff
@@ -0,0 +1,73 @@
+  type: "Assignment" | "Mid Exam" | "Quiz" | "Project" | "Practical Lab" | "Final Exam";
+  weight: number;
+  dueDate: Date;
+};
```
used the global evaluation service

Roheeth05 commented on src/EvaluationCalendar/MonthlyView/MonthlyView.test.tsx:
```diff
@@ -0,0 +1,47 @@
+  {
+    course: "CS101",
+    title: "Assignment 1",
+    type: "Assignment" as const,
```
Resolved

achow-conestogac commented on src/EvaluationCalendar/MonthlyView/MonthlyView.tsx:
```diff
@@ -0,0 +1,67 @@
+  const calendarCells = useMemo(() => {
+    const cells: { date: Date; evaluations: Evaluation[] }[] = [];
+
+    // Add placeholders before the first day
+    for (let i = 0; i < firstWeekday; i++) {
+      cells.push({ date: new Date(NaN), evaluations: [] });
+    }
+
+    // Populate actual days of the month
```
you do not need these comments. remove both.

achow-conestogac reviewed (commented):
You need to put the appropriate index.ts files.

Roheeth05 commented:
Hi Professor, I've integrated MonthlyView into CalendarView

- Added MonthlyView component rendering logic to CalendarView
- Preserved CalendarNavigation for both views
- Fallback message handled for both views when no evaluations are scheduled
- Updated test cases

achow-conestogac commented on src/Evaluation/EvaluationCalendar/CalendarView/CalendarView.tsx:
```diff
@@ -43,41 +44,59 @@ const CalendarView: React.FC<CalendarViewProps> = ({ evaluations }) => {
-          view === "weekly"
-            ? navigateWeek("prev")
-            : navigateMonth("prev")
+          view === "weekly" ? navigateWeek("prev") : navigateMonth("prev")
```
please don't modify code that you don't need to.

it makes the review extremely difficult.

Please remove everything that is not required for this change.

Roheeth05 replied:
Thank you for the feedback Professor. I understand the importance of keeping changes focused and minimal. I've removed unrelated modifications and only included what's necessary for the specific change.

Roheeth05 commented on src/Evaluation/EvaluationCalendar/CalendarView/CalendarView.tsx:
```diff
@@ -43,41 +45,53 @@ const CalendarView: React.FC<CalendarViewProps> = ({ evaluations }) => {
+        />
       </div>
+
+      {showNoEvaluationsMessage ? (
+        <p className="text-center text-gray-500 italic">
+          No evaluations scheduled
+        </p>
+      ) : view === "weekly" ? (
+        <div className="space-y-4">
+          {sortedDates.map((dateStr) => (
+            <CalendarDayCard
+              key={dateStr}
+              date={dateStr}
+              evaluations={groupedByDate[dateStr]}
+            />
+          ))}
+        </div>
+      ) : (
+        <MonthlyView evaluations={evaluations} year={year} month={month} />
+      )}
```
Hi Professor, thank you for the feedback. I've applied the same fix now.

achow-conestogac commented on src/Evaluation/EvaluationCalendar/index.ts:
```diff
@@ -1 +1,3 @@
-export { default as CalendarView } from "./CalendarView"
\ No newline at end of file
+export { default as MonthlyView } from "./MonthlyView/MonthlyView"
```
you shouldn't need to export this.  it's only used on `CalendarView` which is already exported.

achow-conestogac reviewed (commented):
some minor comments and you have conflicts to resolve.

achow-conestogac reviewed (approved):
okay. this looks okay.

please resolve all conversations before merging.
